<a href="https://colab.research.google.com/github/SUKstudent/Infotact_Data_Analytics_Internship/blob/lonali-ray/notebooks/EDA.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

Data Ingestion and Eda
--------------------------------
Importing libraries
*   Panda was used for data loading and manipulation
*   Numpy was used for numerical operation
*   Matplotlib and Seaborn were used for data visualization.



In [2]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns

print("Libraries loaded successfully")

Libraries loaded successfully


Uploading the Dataset
-----------------------------


*   To upload the CSV dataset into google.colab



In [3]:

from google.colab import files
uploaded = files.upload()

Saving multi_touch_attribution_dataset.csv to multi_touch_attribution_dataset.csv


Loading the Dataset
--------------------------


*   To read the CSVfile and store it in a data frame named df
*   Dataframe is the main structure used for analysis



In [7]:
df= pd.read_csv('multi_touch_attribution_dataset.csv')

View First Records
--------------------------
*   To understand the structure of the dataset.
*   To inspect column name and sample values.



In [8]:
df.head()

,event_id,user_id,journey_id,session_id,event_timestamp_utc,event_date,touchpoint_number,total_touchpoints_in_journey,channel,channel_group,...,landing_page,device,region,product_plan,ad_spend,is_conversion,conversion_id,conversion_timestamp_utc,conversion_value,customer_status
0,E000001,U00001,J00001,S00001_01,2026-02-05 11:14:00,2026-02-05,1,4,TikTok Ads,paid_social,...,/free-trial,Mobile,South,Professional,0.20,0,NaN,NaN,0.00,NaN
1,E000002,U00001,J00001,S00001_02,2026-02-07 02:26:00,2026-02-07,2,4,Organic Search,organic,...,/demo,Mobile,South,Professional,0.00,0,NaN,NaN,0.00,NaN
2,E000003,U00001,J00001,S00001_03,2026-02-08 00:17:00,2026-02-08,3,4,Direct,direct,...,/blog/marketing-roi,Mobile,South,Professional,0.00,0,NaN,NaN,0.00,NaN
3,E000004,U00001,J00001,S00001_04,2026-02-08 18:05:00,2026-02-08,4,4,Meta Ads,paid_social,...,/blog/marketing-roi,Mobile,South,Professional,0.62,1,C00001,2026-02-08 18:05:00,267.15,returning
4,E000005,U00002,J00002,S00002_01,2026-03-10 09:59:00,2026-03-10,1,5,LinkedIn Ads,paid_social,...,/pricing,Desktop,West,Starter,3.75,0,NaN,NaN,0.00,NaN


Check Dataset Size
----------------------
1926 records
26 columns



In [9]:
df.shape

(1926, 26)

In [10]:
df.columns

Index(['event_id', 'user_id', 'journey_id', 'session_id',
       'event_timestamp_utc', 'event_date', 'touchpoint_number',
       'total_touchpoints_in_journey', 'channel', 'channel_group', 'campaign',
       'funnel_stage', 'event_type', 'utm_source', 'utm_medium',
       'utm_campaign', 'landing_page', 'device', 'region', 'product_plan',
       'ad_spend', 'is_conversion', 'conversion_id',
       'conversion_timestamp_utc', 'conversion_value', 'customer_status'],
      dtype='object')

In [22]:
df["event_timestamp_utc"] = pd.to_datetime(df["event_timestamp_utc"])
df["event_date"] = pd.to_datetime(df["event_date"])
df["conversion_timestamp_utc"] = pd.to_datetime(df["conversion_timestamp_utc"], errors="coerce")

df["ad_spend"] = pd.to_numeric(df["ad_spend"], errors="coerce").fillna(0)
df["conversion_value"] = pd.to_numeric(df["conversion_value"], errors="coerce").fillna(0)

df = df.drop_duplicates(subset=["event_id"])
df = df.sort_values(
    ["journey_id", "event_timestamp_utc"]
)

In [23]:
df.head()

,event_id,user_id,journey_id,session_id,event_timestamp_utc,event_date,touchpoint_number,total_touchpoints_in_journey,channel,channel_group,...,landing_page,device,region,product_plan,ad_spend,is_conversion,conversion_id,conversion_timestamp_utc,conversion_value,customer_status
0,E000001,U00001,J00001,S00001_01,2026-02-05 11:14:00,2026-02-05,1,4,TikTok Ads,paid_social,...,/free-trial,Mobile,South,Professional,0.20,0,NaN,NaT,0.00,NaN
1,E000002,U00001,J00001,S00001_02,2026-02-07 02:26:00,2026-02-07,2,4,Organic Search,organic,...,/demo,Mobile,South,Professional,0.00,0,NaN,NaT,0.00,NaN
2,E000003,U00001,J00001,S00001_03,2026-02-08 00:17:00,2026-02-08,3,4,Direct,direct,...,/blog/marketing-roi,Mobile,South,Professional,0.00,0,NaN,NaT,0.00,NaN
3,E000004,U00001,J00001,S00001_04,2026-02-08 18:05:00,2026-02-08,4,4,Meta Ads,paid_social,...,/blog/marketing-roi,Mobile,South,Professional,0.62,1,C00001,2026-02-08 18:05:00,267.15,returning
4,E000005,U00002,J00002,S00002_01,2026-03-10 09:59:00,2026-03-10,1,5,LinkedIn Ads,paid_social,...,/pricing,Desktop,West,Starter,3.75,0,NaN,NaT,0.00,NaN


In [25]:
df.to_csv("cleaned_mta_roi_dataset.csv", index=False)

In [33]:
#check how many unique cusomer
df['user_id'].nunique()

500

In [34]:
#Check no of unique journey
df['journey_id'].nunique()

500

In [22]:

#Check no of unique chanel
df['channel'].nunique()

8

Calculated key marketing performance metrics including Total Ad Spend, Total Revenue, Total Conversions, and Total Customer Journeys. Using these metrics, I derived ROAS to measure advertising efficiency, CAC to evaluate customer acquisition costs, and Conversion Rate to assess the effectiveness of marketing efforts in converting customer journeys into purchases. These KPIs provide a quantitative view of overall marketing performance and help support budget allocation decisions.

In [27]:
total_spend = df["ad_spend"].sum()
total_revenue = df["conversion_value"].sum()
total_conversions = df["is_conversion"].sum()
total_journeys = df["journey_id"].nunique()

roas = total_revenue / total_spend
cac = total_spend / total_conversions
conversion_rate = total_conversions / total_journeys

print("Total Spend:", round(total_spend, 2))
print("Total Revenue:", round(total_revenue, 2))
print("Total Conversions:", total_conversions)
print("Total Journeys:", total_journeys)
print("ROAS:", round(roas, 2))
print("CAC:", round(cac, 2))
print("Conversion Rate:", round(conversion_rate * 100, 2), "%")


Total Spend: 1920.94
Total Revenue: 61086.75
Total Conversions: 209
Total Journeys: 500
ROAS: 31.8
CAC: 9.19
Conversion Rate: 41.8 %


In [23]:
#check unique chanel
df['channel'].unique()

array(['TikTok Ads', 'Organic Search', 'Direct', 'Meta Ads',
       'LinkedIn Ads', 'Email', 'Google Ads', 'Referral'], dtype=object)

The dataset contains 500 unique customers and 500 unique customer journeys, indicating a one-to-one relationship between customers and journeys. Eight distinct marketing channels are represented, including paid advertising, organic traffic, referral sources, and email marketing.

In [25]:
df['channel'].value_counts()

,count
channel,
Organic Search,361
Google Ads,328
Meta Ads,303
LinkedIn Ads,222
Email,202
Direct,185
TikTok Ads,177
Referral,148


In [26]:
df['funnel_stage'].value_counts()

,count
funnel_stage,
Consideration,684
Decision,560
Awareness,473
Purchase,209


In [27]:
df['is_conversion'].sum()

np.int64(209)

In [28]:
df['ad_spend'].sum()

np.float64(1920.94)

In [29]:
df['conversion_value'].sum()

np.float64(61086.75)

In [30]:
roas = df['conversion_value'].sum() / df['ad_spend'].sum()
roas

np.float64(31.800446656324507)

In [31]:
df.groupby('channel')['conversion_value'].sum().sort_values(ascending=False)

,conversion_value
channel,
Google Ads,19152.57
Direct,15021.46
Email,12322.49
Organic Search,7363.02
LinkedIn Ads,4825.41
TikTok Ads,1081.29
Referral,722.89
Meta Ads,597.62


In [32]:
df.groupby('channel')['ad_spend'].sum().sort_values(ascending=False)

,ad_spend
channel,
Google Ads,723.08
LinkedIn Ads,685.33
Meta Ads,401.32
TikTok Ads,101.07
Email,10.14
Direct,0.00
Organic Search,0.00
Referral,0.00


In [35]:
channel_performance = df.groupby('channel').agg({
    'ad_spend':'sum',
    'conversion_value':'sum'
})

channel_performance['ROAS'] = (
    channel_performance['conversion_value']
    / channel_performance['ad_spend']
)

channel_performance

,ad_spend,conversion_value,ROAS
channel,,,
Direct,0.00,15021.46,inf
Email,10.14,12322.49,1215.235700
Google Ads,723.08,19152.57,26.487484
LinkedIn Ads,685.33,4825.41,7.041002
Meta Ads,401.32,597.62,1.489136
Organic Search,0.00,7363.02,inf
Referral,0.00,722.89,inf
TikTok Ads,101.07,1081.29,10.698427


In [36]:
df.groupby('channel')['is_conversion'].sum().sort_values(ascending=False)

,is_conversion
channel,
Google Ads,67
Direct,48
Email,40
Organic Search,29
LinkedIn Ads,13
Referral,5
Meta Ads,4
TikTok Ads,3


In [37]:
print("Total Records:", len(df))
print("Unique Customers:", df['user_id'].nunique())
print("Unique Journeys:", df['journey_id'].nunique())
print("Total Conversions:", df['is_conversion'].sum())
print("Total Revenue:", df['conversion_value'].sum())
print("Total Ad Spend:", df['ad_spend'].sum())
print("ROAS:", round(df['conversion_value'].sum()/df['ad_spend'].sum(),2))

Total Records: 1926
Unique Customers: 500
Unique Journeys: 500
Total Conversions: 209
Total Revenue: 61086.75
Total Ad Spend: 1920.94
ROAS: 31.8


In [28]:
channel_summary = df.groupby("channel").agg(
    total_spend=("ad_spend", "sum"),
    total_revenue=("conversion_value", "sum"),
    conversions=("is_conversion", "sum"),
    touchpoints=("event_id", "count")
).reset_index()

channel_summary["roas"] = channel_summary["total_revenue"] / channel_summary["total_spend"]
channel_summary["cac"] = channel_summary["total_spend"] / channel_summary["conversions"]

channel_summary = channel_summary.replace([float("inf"), -float("inf")], 0)
channel_summary = channel_summary.fillna(0)

channel_summary

,channel,total_spend,total_revenue,conversions,touchpoints,roas,cac
0,Direct,0.00,15021.46,48,185,0.000000,0.000000
1,Email,10.14,12322.49,40,202,1215.235700,0.253500
2,Google Ads,723.08,19152.57,67,328,26.487484,10.792239
3,LinkedIn Ads,685.33,4825.41,13,222,7.041002,52.717692
4,Meta Ads,401.32,597.62,4,303,1.489136,100.330000
5,Organic Search,0.00,7363.02,29,361,0.000000,0.000000
6,Referral,0.00,722.89,5,148,0.000000,0.000000
7,TikTok Ads,101.07,1081.29,3,177,10.698427,33.690000


# Week 2 - SQL Attribution Analysis
#In this section, the dataset is loaded into SQLite and analyzed using SQL queries. Window functions are used to sequence user journeys and perform attribution analysis.

In [29]:
import sqlite3

In [38]:
import sqlite3

conn = sqlite3.connect('marketing.db')

df.to_sql(
    'marketing_data',
    conn,
    if_exists='replace',
    index=False
)

1926

In [50]:
query = """
SELECT *
FROM marketing_data
LIMIT 5
"""
pd.read_sql_query(query, conn)

,event_id,user_id,journey_id,session_id,event_timestamp_utc,event_date,touchpoint_number,total_touchpoints_in_journey,channel,channel_group,...,landing_page,device,region,product_plan,ad_spend,is_conversion,conversion_id,conversion_timestamp_utc,conversion_value,customer_status
0,E000001,U00001,J00001,S00001_01,2026-02-05 11:14:00,2026-02-05,1,4,TikTok Ads,paid_social,...,/free-trial,Mobile,South,Professional,0.20,0,None,None,0.00,None
1,E000002,U00001,J00001,S00001_02,2026-02-07 02:26:00,2026-02-07,2,4,Organic Search,organic,...,/demo,Mobile,South,Professional,0.00,0,None,None,0.00,None
2,E000003,U00001,J00001,S00001_03,2026-02-08 00:17:00,2026-02-08,3,4,Direct,direct,...,/blog/marketing-roi,Mobile,South,Professional,0.00,0,None,None,0.00,None
3,E000004,U00001,J00001,S00001_04,2026-02-08 18:05:00,2026-02-08,4,4,Meta Ads,paid_social,...,/blog/marketing-roi,Mobile,South,Professional,0.62,1,C00001,2026-02-08 18:05:00,267.15,returning
4,E000005,U00002,J00002,S00002_01,2026-03-10 09:59:00,2026-03-10,1,5,LinkedIn Ads,paid_social,...,/pricing,Desktop,West,Starter,3.75,0,None,None,0.00,None


In [53]:
query = """
SELECT
    user_id,
    journey_id,
    channel,
    event_timestamp_utc,

    ROW_NUMBER() OVER(
        PARTITION BY user_id
        ORDER BY event_timestamp_utc
    ) AS touch_order

FROM marketing_data
"""

journey_sequence = pd.read_sql_query(query, conn)

journey_sequence.head(20)

,user_id,journey_id,channel,event_timestamp_utc,touch_order
0,U00001,J00001,TikTok Ads,2026-02-05 11:14:00,1
1,U00001,J00001,Organic Search,2026-02-07 02:26:00,2
2,U00001,J00001,Direct,2026-02-08 00:17:00,3
3,U00001,J00001,Meta Ads,2026-02-08 18:05:00,4
4,U00002,J00002,LinkedIn Ads,2026-03-10 09:59:00,1
5,U00002,J00002,Direct,2026-03-11 08:04:00,2
6,U00002,J00002,TikTok Ads,2026-03-12 11:14:00,3
7,U00002,J00002,Meta Ads,2026-03-14 05:54:00,4
8,U00002,J00002,Email,2026-03-16 02:38:00,5
9,U00003,J00003,Email,2026-02-04 09:13:00,1


In [54]:
query = """
WITH first_touch AS (

    SELECT
        user_id,
        channel,

        ROW_NUMBER() OVER(
            PARTITION BY user_id
            ORDER BY event_timestamp_utc
        ) AS touch_order

    FROM marketing_data

)

SELECT
    channel,
    COUNT(*) AS first_touch_count

FROM first_touch

WHERE touch_order = 1

GROUP BY channel

ORDER BY first_touch_count DESC
"""

first_touch_result = pd.read_sql_query(query, conn)

first_touch_result

,channel,first_touch_count
0,Meta Ads,117
1,TikTok Ads,102
2,Organic Search,85
3,Google Ads,58
4,LinkedIn Ads,54
5,Referral,42
6,Direct,26
7,Email,16


First-Touch Attribution analysis revealed that Meta Ads and TikTok Ads are the primary customer acquisition channels, contributing 117 and 102 first interactions respectively. Although Google Ads generated the highest conversion volume and revenue, it accounted for only 58 first-touch interactions, suggesting that it plays a stronger role in conversion rather than initial customer acquisition.

In [57]:
query = """
SELECT
    user_id,
    channel,

    COUNT(*) OVER(
        PARTITION BY user_id
    ) AS total_touchpoints,

    1.0 /
    COUNT(*) OVER(
        PARTITION BY user_id
    ) AS linear_weight

FROM marketing_data
"""

linear_weights = pd.read_sql_query(query, conn)

linear_weights.head(20)

,user_id,channel,total_touchpoints,linear_weight
0,U00001,TikTok Ads,4,0.25
1,U00001,Organic Search,4,0.25
2,U00001,Direct,4,0.25
3,U00001,Meta Ads,4,0.25
4,U00002,LinkedIn Ads,5,0.20
5,U00002,Direct,5,0.20
6,U00002,TikTok Ads,5,0.20
7,U00002,Meta Ads,5,0.20
8,U00002,Email,5,0.20
9,U00003,Email,1,1.00


In [56]:
query = """
WITH linear_attr AS (

SELECT
    channel,

    1.0 /
    COUNT(*) OVER(
        PARTITION BY user_id
    ) AS linear_weight

FROM marketing_data

)

SELECT
    channel,
    ROUND(SUM(linear_weight),2) AS attributed_credit

FROM linear_attr

GROUP BY channel

ORDER BY attributed_credit DESC
"""

linear_attribution = pd.read_sql_query(query, conn)

linear_attribution

,channel,attributed_credit
0,Organic Search,90.89
1,Google Ads,84.82
2,Meta Ads,79.26
3,LinkedIn Ads,58.15
4,Email,50.30
5,TikTok Ads,49.67
6,Direct,47.62
7,Referral,39.28
